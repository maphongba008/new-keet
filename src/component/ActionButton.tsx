import React from 'react'
import { StyleSheet, Text, View } from 'react-native'

import { ButtonBase } from 'component/Button'
import SvgIcon, { SvgIconType } from 'component/SvgIcon'
import { colors, createThemedStylesheet } from 'component/theme'
import s, { UI_SIZE_8, UI_SIZE_12, UI_SIZE_18 } from 'lib/commonStyles'

interface ActionButtonI {
  label: string
  iconLeft?: SvgIconType | React.ReactNode
  iconRight?: SvgIconType | React.ReactNode
  onPressItem: () => void
}

function ActionButton(props: ActionButtonI) {
  const { label, iconLeft, iconRight, onPressItem, ...rest } = props
  const styles = getStyles()

  return (
    <>
      <ButtonBase style={styles.actionBtn} onPress={onPressItem} {...rest}>
        <View style={[s.row, s.alignItemsEnd]}>
          {typeof iconLeft === 'string' ? (
            <SvgIcon
              name={iconLeft as SvgIconType}
              width={UI_SIZE_18}
              height={UI_SIZE_18}
              color={colors.white_snow}
            />
          ) : (
            iconLeft
          )}
          <View style={[s.row, s.alignItemsCenter, styles.textWrapper]}>
            <Text style={styles.btnTypeLabel}>{label}</Text>
          </View>
        </View>
        {typeof iconRight === 'string' ? (
          <SvgIcon
            name={iconRight as SvgIconType}
            width={UI_SIZE_18}
            height={UI_SIZE_18}
            color={colors.white_snow}
          />
        ) : (
          iconRight
        )}
      </ButtonBase>
    </>
  )
}

const getStyles = createThemedStylesheet((theme) => {
  const styles = StyleSheet.create({
    actionBtn: {
      ...s.row,
      ...s.alignItemsCenter,
      ...s.flexSpaceBetween,
      backgroundColor: theme.color.grey_600,
      borderRadius: UI_SIZE_12,
      height: 52,
      marginTop: theme.spacing.normal,
      padding: theme.spacing.standard,
    },
    btnTypeLabel: {
      ...theme.text.body,
      fontSize: 15,
    },
    textWrapper: {
      marginLeft: UI_SIZE_8,
    },
  })
  return styles
})

export default ActionButton
