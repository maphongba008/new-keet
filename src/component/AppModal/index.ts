import AppModal from './AppModal'
import * as ModalTypes from './types'

export default AppModal
export { ModalTypes }
