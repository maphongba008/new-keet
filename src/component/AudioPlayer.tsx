import React, {
  memo,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react'
import { StyleSheet, Text, View } from 'react-native'
import { useDispatch, useSelector } from 'react-redux'
import {
  Audio,
  AVPlaybackStatus,
  InterruptionModeAndroid,
  InterruptionModeIOS,
} from 'expo-av'
import isEqual from 'react-fast-compare'
import { useIsFocused } from '@react-navigation/native'
import _isEmpty from 'lodash/isEmpty'

import { getCurrentSoundUri, setSoundUri } from 'reducers/application'

import s, {
  ICON_SIZE_32,
  UI_SIZE_2,
  UI_SIZE_4,
  UI_SIZE_8,
  UI_SIZE_16,
  UI_SIZE_20,
  UI_SIZE_24,
  UI_SIZE_32,
  UI_SIZE_120,
} from 'lib/commonStyles'
import {
  ONE_HOUR_IN_MILLISECONDS,
  ONE_MINUTE_IN_MILLISECONDS,
  ONE_SECOND_IN_MILLISECONDS,
} from 'lib/constants'
import useStateDeepEqual from 'lib/hooks/useStateDeepEqual'
import { isAndroid } from 'lib/platform'
import { getMessageCellAvailableWidth } from 'lib/size'

import { useAppBottomSheetState } from './AppBottomSheet/AppBottomSheet.Store'
import BottomSheetEnum from './AppBottomSheet/SheetComponents/BottomSheetEnum'
import { IconButton } from './Button'
import { Loading } from './Loading'
import Slider from './Slider'
import SvgIcon from './SvgIcon'
import { colors, createThemedStylesheet } from './theme'
import WaveformVisual from './WaveformVisual'

type PlayerStatus = {
  durationMillis?: number
  isPlaying?: boolean
  positionMillis?: number
  isBuffering?: boolean
}

type AudioPlayerProps = {
  showShowSlider?: boolean
  onAudioLoaded?: Function
  uri: string
  fromChat?: boolean
  fromAttachment?: boolean
  loading?: boolean
  samples?: number[]
}

export const millisecondsToAudioDurationFormat = (ms?: number) => {
  if (!ms) return '0:00'

  let msLeft = ms
  const hours = Math.floor(msLeft / ONE_HOUR_IN_MILLISECONDS)
  msLeft -= hours * ONE_HOUR_IN_MILLISECONDS
  const minutes = Math.floor(msLeft / ONE_MINUTE_IN_MILLISECONDS)
  msLeft -= minutes * ONE_MINUTE_IN_MILLISECONDS
  const seconds = Math.floor(msLeft / ONE_SECOND_IN_MILLISECONDS)

  const minutesString = String(minutes).padStart(2, '0')
  const secondsString = String(seconds).padStart(2, '0')

  if (!hours) return `${minutes}:${secondsString}`

  return `${hours}:${minutesString}:${secondsString}`
}

const defaultDuration = millisecondsToAudioDurationFormat()

const PLAYSPEED_OPTIONS = {
  half: 0.5,
  normal: 1,
  double: 2,
}

const nextPlaySpeed = (ps: number) => {
  if (ps === PLAYSPEED_OPTIONS.half) return PLAYSPEED_OPTIONS.normal
  if (ps === PLAYSPEED_OPTIONS.normal) return PLAYSPEED_OPTIONS.double
  return PLAYSPEED_OPTIONS.half
}

const AudioPlayer = ({
  uri,
  fromChat,
  onAudioLoaded,
  fromAttachment,
  showShowSlider = true,
  loading = false,
  samples = [],
}: AudioPlayerProps) => {
  const styles = getStyles()

  const dispatch = useDispatch()
  const currentSoundUri = useSelector(getCurrentSoundUri)

  const isFocused = useIsFocused()
  const { showBottomSheet, sheetType } = useAppBottomSheetState()

  const [playback, setPlayback] = useStateDeepEqual<PlayerStatus>()
  const [sound, setSound] = useState<Audio.Sound>()
  const [sliderValue, setSliderValue] = useState(0)
  const [wasPlaying, setWasPlaying] = useState(false)
  const [playSpeed, setPlaySpeed] = useState(PLAYSPEED_OPTIONS.normal)

  const duration = useRef<string>(defaultDuration)

  const isVoiceNote = !_isEmpty(samples)
  const iconSize = fromAttachment ? UI_SIZE_24 : UI_SIZE_32 - UI_SIZE_2

  useEffect((): any => {
    return async () => {
      return sound ? await sound?.unloadAsync() : undefined
    }
  }, [isFocused, sound])

  const initSoundMode = useCallback(
    async () =>
      await Audio.setAudioModeAsync({
        playsInSilentModeIOS: true,
        shouldDuckAndroid: true,
        interruptionModeAndroid: InterruptionModeAndroid.DoNotMix,
        interruptionModeIOS: InterruptionModeIOS.DoNotMix,
      }),
    [],
  )

  const handleDuration = useCallback((millis?: number) => {
    const durationTimestamp = millisecondsToAudioDurationFormat(millis)
    duration.current = durationTimestamp
  }, [])

  const totalDuration = useMemo(
    () => millisecondsToAudioDurationFormat(playback?.durationMillis),
    [playback?.durationMillis],
  )

  const isBuffering = useMemo(
    () => (playback?.isPlaying && playback?.isBuffering) || false,
    [playback?.isBuffering, playback?.isPlaying],
  )

  const unloadAudioFromMemory = useCallback(async () => {
    if (!sound) return
    await sound?.unloadAsync()
    setSound(undefined)
  }, [sound])

  const stopSound = useCallback(async () => {
    await sound?.stopAsync()
    duration.current = defaultDuration
    setSliderValue(0)
    unloadAudioFromMemory()
  }, [sound, unloadAudioFromMemory])

  useEffect(() => {
    if (
      (totalDuration !== defaultDuration || playback?.durationMillis !== 0) &&
      playback?.positionMillis === playback?.durationMillis
    ) {
      const timeout = setTimeout(() => stopSound(), 200)
      return () => clearTimeout(timeout)
    }
  }, [
    playback?.durationMillis,
    playback?.positionMillis,
    stopSound,
    totalDuration,
  ])

  const playbackStatusUpdate = useCallback(
    async (playStatus: AVPlaybackStatus) => {
      if (!playStatus.isLoaded) {
        if (playStatus.error) {
          console.log(
            `Encountered a fatal error during playback: ${playStatus.error}`,
          )
        }
        return
      }
      setPlayback({
        isPlaying: playStatus.isPlaying,
        positionMillis: playStatus.positionMillis,
        durationMillis: playStatus.durationMillis,
        isBuffering: playStatus.isBuffering,
      })
      if (wasPlaying && playStatus.positionMillis === 0) {
        setWasPlaying(false)
        return
      }
      handleDuration(playStatus?.positionMillis)
      setSliderValue(playStatus?.positionMillis)
    },
    [setPlayback, wasPlaying, handleDuration],
  )

  useEffect((): any => {
    const initSound = async () => {
      try {
        const soundObj = new Audio.Sound()
        const audio = await soundObj.loadAsync({ uri })
        if (audio.isLoaded) {
          onAudioLoaded?.()
        }
        const playStatus = await soundObj.getStatusAsync()
        // @ts-ignore
        playbackStatusUpdate({ ...playStatus, isBuffering: false }) // Buffering will be handled while play
        // Init just to load the audio meta like duration, So safe to unload from memory. Will load again while play
        await soundObj?.unloadAsync()
      } catch (e) {
        console.log('initSound Error', e)
      }
    }
    if (!isFocused) return
    initSound()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [uri, isFocused])

  const playAudio = useCallback(async () => {
    try {
      if (!uri) return
      let _sound: Audio.Sound | null = null
      if (!sound) {
        const { sound: soundObj } = await Audio.Sound.createAsync({ uri })
        _sound = soundObj
        setSound(soundObj)
      } else {
        // @ts-ignore
        _sound = sound
      }
      _sound?.setProgressUpdateIntervalAsync(isAndroid ? 200 : 150)
      _sound?.setOnPlaybackStatusUpdate(playbackStatusUpdate)
      _sound?.setStatusAsync({ rate: playSpeed })
      await initSoundMode()
      await _sound?.playFromPositionAsync(sliderValue)
      dispatch(setSoundUri(uri))
    } catch (e) {
      console.log('playAudio error', e)
    }
  }, [
    dispatch,
    initSoundMode,
    playSpeed,
    playbackStatusUpdate,
    sliderValue,
    sound,
    uri,
  ])

  const pauseAndUnloadAudio = useCallback(async () => {
    await sound?.pauseAsync()
    await unloadAudioFromMemory()
    setWasPlaying(true)
  }, [sound, unloadAudioFromMemory])

  useEffect(() => {
    // Pause audio if option sheet is opened on android to avoid core issue
    if (!sound || !playback || !isAndroid) return
    if (
      playback?.isPlaying &&
      showBottomSheet &&
      sheetType === BottomSheetEnum.OptionsSheet
    ) {
      pauseAndUnloadAudio()
    }
  }, [pauseAndUnloadAudio, playback, sheetType, showBottomSheet, sound])

  useEffect(() => {
    if (
      uri !== currentSoundUri &&
      currentSoundUri !== '' &&
      playback?.isPlaying
    ) {
      pauseAndUnloadAudio()
    }
  }, [currentSoundUri, pauseAndUnloadAudio, playback?.isPlaying, uri])

  const handleOnValueChange = useCallback(
    (_sliderValue: number) => {
      const value = Math.floor(_sliderValue)
      setSliderValue(value)
      handleDuration(value)
    },
    [handleDuration],
  )

  const handleOnSlidingStart = useCallback(async () => {
    setWasPlaying(true)
    if (!playback?.isPlaying) return
    await sound?.pauseAsync()
  }, [playback?.isPlaying, sound])

  const handleOnSlidingComplete = useCallback(
    async (_sliderValue: number) => {
      try {
        const value = Math.floor(_sliderValue)
        const status = await sound?.setPositionAsync(value)
        if (!status) return
        await sound?.setStatusAsync(status)
        if (wasPlaying) {
          await playAudio()
          setWasPlaying(false)
        }
      } catch (e) {
        console.log('slider change error', e)
      }
    },
    [playAudio, sound, wasPlaying],
  )

  const onTogglePlaybackSpeed = useCallback(async () => {
    try {
      const nextSpeed = nextPlaySpeed(playSpeed)
      setPlaySpeed(nextSpeed)
      sound?.setStatusAsync({ rate: nextSpeed })
    } catch (e) {
      console.log('toggle playback speed change error', e)
    }
  }, [playSpeed, sound])

  const containerStyle = useMemo(
    () => ({
      ...(fromAttachment
        ? styles.attachmentPlayer
        : { paddingVertical: isVoiceNote ? UI_SIZE_8 : undefined }),
    }),
    [fromAttachment, styles.attachmentPlayer, isVoiceNote],
  )

  const waveformWidth = useMemo(() => {
    const availableWidth = getMessageCellAvailableWidth()

    if (fromChat) {
      if (isVoiceNote) {
        return availableWidth - UI_SIZE_120 - UI_SIZE_20
      }
      return availableWidth - UI_SIZE_120 // extra space for mute icon
    }

    return availableWidth - (UI_SIZE_120 - UI_SIZE_20) // -100
  }, [fromChat, isVoiceNote])

  const SliderComponent = useCallback(() => {
    if (!showShowSlider) return null

    return (
      <Slider
        style={styles.slider}
        disabled={isBuffering}
        onValueChange={handleOnValueChange}
        value={sliderValue}
        minimumValue={0}
        maximumValue={playback?.durationMillis}
        onSlidingStart={handleOnSlidingStart}
        onSlidingComplete={handleOnSlidingComplete}
        minimumTrackTintColor={_isEmpty(samples) ? null : 'transparent'}
        maximumTrackTintColor={_isEmpty(samples) ? null : 'transparent'}
        offset={0}
        thumbVisible={!isVoiceNote}
      />
    )
  }, [
    handleOnSlidingComplete,
    handleOnSlidingStart,
    handleOnValueChange,
    isBuffering,
    playback?.durationMillis,
    samples,
    showShowSlider,
    sliderValue,
    styles.slider,
    isVoiceNote,
  ])

  if (loading) return null

  return (
    <View
      style={[
        s.centerAlignedRow,
        !fromChat ? s.container : s.flexSpaceBetween,
        !fromChat && !fromAttachment ? styles.attachmentVoicePlayer : {},
        containerStyle,
      ]}
    >
      <View style={[s.row, styles.playbackAction]}>
        {isBuffering ? (
          <IconButton disabled style={styles.icon}>
            <Loading style={styles.audioBuffering} />
          </IconButton>
        ) : !playback?.isPlaying ? (
          <IconButton onPress={playAudio} style={styles.icon}>
            <SvgIcon
              name="playCircleFilled"
              width={iconSize}
              height={iconSize}
            />
          </IconButton>
        ) : (
          <IconButton onPress={pauseAndUnloadAudio} style={styles.icon}>
            <SvgIcon
              name="pauseCircleFilled"
              width={iconSize}
              height={iconSize}
            />
          </IconButton>
        )}
      </View>
      {fromAttachment && (
        <Text
          style={[
            styles.duration,
            isVoiceNote && styles.durationPartial,
            fromAttachment && styles.durationAttachment,
          ]}
          numberOfLines={1}
          adjustsFontSizeToFit
        >
          {isVoiceNote
            ? `${duration.current}`
            : `${duration.current}/${totalDuration}`}
        </Text>
      )}
      <View style={s.centerAlignedRow}>
        {!fromAttachment && (
          <View style={[styles.waveformContainer, { width: waveformWidth }]}>
            <WaveformVisual
              waveform={samples}
              width={waveformWidth}
              positionMillis={sliderValue}
              durationMillis={playback?.durationMillis || 0}
            />
            {SliderComponent()}
          </View>
        )}
        {isVoiceNote && (
          <Text
            style={[styles.duration, styles.durationPartial]}
            numberOfLines={1}
            adjustsFontSizeToFit
          >
            {totalDuration}
          </Text>
        )}
      </View>
      {!fromAttachment && (
        <>
          {fromChat && (
            <IconButton
              style={styles.playbackSpeed}
              onPress={onTogglePlaybackSpeed}
            >
              <Text style={styles.playbackSpeedText}>{playSpeed}x</Text>
            </IconButton>
          )}
        </>
      )}
    </View>
  )
}

export default memo(AudioPlayer, isEqual)

const getStyles = createThemedStylesheet((theme) => {
  const styles = StyleSheet.create({
    attachmentPlayer: {
      padding: UI_SIZE_4,
      ...s.centeredLayout,
      ...s.fullHeight,
    },
    attachmentVoicePlayer: {
      justifyContent: 'space-between',
      paddingHorizontal: UI_SIZE_8,
    },
    audioBuffering: {
      height: UI_SIZE_16,
      width: UI_SIZE_16,
    },
    duration: {
      ...s.textAlignCenter,
      ...theme.text.bodySemiBold,
      fontSize: 13,
    },
    durationAttachment: {
      marginHorizontal: UI_SIZE_4,
    },
    durationPartial: {
      width: UI_SIZE_32,
    },
    icon: {
      height: ICON_SIZE_32,
      width: ICON_SIZE_32,
    },
    playbackAction: {
      width: ICON_SIZE_32,
      ...s.alignItemsCenter,
    },
    playbackSpeed: {
      borderColor: colors.white_snow,
      borderRadius: UI_SIZE_32,
      borderWidth: 1,
      height: UI_SIZE_24,
      paddingVertical: UI_SIZE_2,
      width: 38,
    },
    playbackSpeedText: {
      ...theme.text.bodySemiBold,
      fontSize: 13,
    },
    slider: {
      position: 'absolute',
      zIndex: 1,
      ...s.fullWidth,
      minHeight: 40,
    },
    waveformContainer: {
      justifyContent: 'center',
      marginHorizontal: UI_SIZE_4,
    },
  })
  return styles
})
