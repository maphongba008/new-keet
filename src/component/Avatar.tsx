import React, { memo, ReactNode, useMemo } from 'react'
import {
  ImageStyle,
  StyleProp,
  StyleSheet,
  Text,
  View,
  ViewProps,
  ViewStyle,
} from 'react-native'
import { Image } from 'expo-image'

import { getFirstLetters } from '@holepunchto/keet-store/store/lib/string'

import { APPIUM_IDs, appiumTestProps } from 'lib/appium'
import { getScaledFontSize, randomColor } from 'lib/avatar'
import s, {
  ICON_SIZE_72,
  UI_SIZE_2,
  UI_SIZE_12,
  UI_SIZE_32,
} from 'lib/commonStyles'
import { MemberType } from 'lib/types'

import SvgIcon from './SvgIcon'
import { colors, createThemedStylesheet } from './theme'

// https://github.com/holepunchto/settings/blob/651dcd959994fc1f270f2da5ffbf37fb3fa0ba2e/settings.js#L150

export interface AvatarProps {
  base64?: string | null
  seed?: string
  small?: boolean
  style?: ViewProps['style']
  name?: string | null
  color?: string | null
}

export interface AvatarContainerProps {
  children?: ReactNode
  small?: boolean
  style?: ViewProps['style']
}

export const AvatarContainer: React.FC<AvatarContainerProps> = ({
  small,
  style,
  children,
}) => {
  const styles = getStyles()
  const containerStyle = useMemo(() => {
    if (small) {
      return [style, styles.smallContainer]
    }

    return [style]
  }, [small, style, styles.smallContainer])

  return <View style={containerStyle}>{children}</View>
}

export const Avatar: React.FC<AvatarProps> = memo(
  ({ base64: uri, seed, small, style, color, name }) => {
    const styles = getStyles()

    const avatarDisplayName = getFirstLetters(name ?? '').substring(0, 2)
    const avatarStyle = useMemo(() => {
      const avatarStyles: StyleProp<ViewStyle> = []

      if (!uri) {
        avatarStyles.push(s.centeredLayout, {
          backgroundColor: color || randomColor(seed),
        })
      }

      if (small) {
        avatarStyles.push(styles.smallImg, style)
      } else {
        avatarStyles.push(styles.bigImg, style)
      }

      return StyleSheet.flatten(avatarStyles)
    }, [color, seed, small, style, styles.bigImg, styles.smallImg, uri])

    const circleSize = small ? 16 : 40

    if (uri) {
      return (
        <Image
          style={avatarStyle as ImageStyle}
          source={{ uri }}
          contentFit="contain"
          recyclingKey={uri}
          cachePolicy="memory"
          {...appiumTestProps(APPIUM_IDs.avatar_image)}
        />
      )
    }

    if (avatarDisplayName) {
      return (
        <View style={avatarStyle} {...appiumTestProps(APPIUM_IDs.avatar_image)}>
          <Text
            style={[
              styles.avatarText,
              {
                fontSize: getScaledFontSize(avatarStyle),
              },
            ]}
          >
            {avatarDisplayName}
          </Text>
        </View>
      )
    }
    return (
      <View style={avatarStyle} {...appiumTestProps(APPIUM_IDs.avatar_image)}>
        <SvgIcon
          name="keetOutline"
          width={circleSize}
          height={circleSize}
          color={colors.white_snow}
        />
      </View>
    )
  },
)

type MemberAvatarProps = {
  member: Partial<MemberType>
} & AvatarProps

export const MemberAvatar = ({ member, ...props }: MemberAvatarProps) => {
  const base64 = useMemo(() => {
    return member?.blocked ? null : member?.avatarUrl
  }, [member])

  return (
    <Avatar
      name={member.displayName}
      small
      base64={base64}
      color={member?.color}
      seed={member?.swarmId}
      {...props}
    />
  )
}

export const AVATAR_SIZE = UI_SIZE_32 - UI_SIZE_2
export const AVATAR_SIZE_BIG = ICON_SIZE_72
export const LOBBY_AVATAR_SIZE = 55

interface AvatarStyleSizeParams {
  size: number
  borderWidth: number
  padding: number
}

const getStyles = createThemedStylesheet((theme) => {
  const smallPreset: AvatarStyleSizeParams = {
    size: AVATAR_SIZE,
    borderWidth: 1,
    padding: 1,
  }
  const bigPreset: AvatarStyleSizeParams = {
    size: AVATAR_SIZE_BIG,
    borderWidth: 2,
    padding: 1,
  }

  const getContainerStyle = ({
    size,
    borderWidth,
    padding,
  }: AvatarStyleSizeParams) => ({
    borderRadius: size / 2,
    borderWidth,
    padding,
    borderColor: theme.color.red_400,
    width: size,
    height: size,
  })

  const getImgStyle = (
    { size, borderWidth, padding }: AvatarStyleSizeParams,
    isHighlighted?: boolean,
  ) => {
    const sizeInset = isHighlighted ? size - 2 * (borderWidth + padding) : size

    return {
      width: sizeInset,
      height: sizeInset,
      borderRadius: sizeInset / 2,
    }
  }

  const styles = StyleSheet.create({
    avatarText: {
      ...theme.text.bodyBold,
      fontSize: UI_SIZE_12,
    },
    bigImg: getImgStyle(bigPreset),
    smallContainer: getContainerStyle(smallPreset),
    smallImg: getImgStyle(smallPreset),
  })
  return styles
})
