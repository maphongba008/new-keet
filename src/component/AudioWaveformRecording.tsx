import React, { memo, useCallback, useEffect, useMemo, useRef } from 'react'
import { ScrollView, StyleSheet, View } from 'react-native'
import Reanimated, { ZoomIn } from 'react-native-reanimated'
import _isEmpty from 'lodash/isEmpty'

import { createThemedStylesheet } from 'component/theme'
import s, { UI_SIZE_4, UI_SIZE_40 } from 'lib/commonStyles'
import {
  WAVEFORM_HEIGHT,
  WAVEFORM_MIN_HEIGHT,
  WAVEFORM_SPACE,
} from 'lib/constants'
import { getMessageCellAvailableWidth } from 'lib/size'

const RECORDING_WAVEFORM_WIDTH = 2

type AudioWaveformRecordingProps = {
  samples: Array<number> | undefined
}

const AudioWaveformRecording = ({
  samples = [],
}: AudioWaveformRecordingProps) => {
  const styles = getStyles()
  const scrollRef = useRef<ScrollView>(null)

  const maxSampleCount = useMemo(() => {
    const availableWidth = getMessageCellAvailableWidth() - UI_SIZE_40
    return availableWidth / (RECORDING_WAVEFORM_WIDTH + WAVEFORM_SPACE * 2)
  }, [])

  useEffect(() => {
    if (samples.length % 2 !== 0 || samples.length < maxSampleCount) return

    scrollRef?.current?.scrollToEnd({ animated: true })
  }, [maxSampleCount, samples.length])

  const renderWaveform = useCallback(
    (sample: number, key: number) => (
      <Reanimated.View
        key={key}
        entering={ZoomIn}
        style={[
          styles.waveformItem,
          {
            height: Math.min(
              Math.max(sample / 10, WAVEFORM_MIN_HEIGHT),
              WAVEFORM_HEIGHT,
            ),
          },
        ]}
      />
    ),
    [styles.waveformItem],
  )

  if (_isEmpty(samples)) {
    return (
      <View style={[s.container, s.centeredLayout]}>
        <View style={styles.hLine} />
      </View>
    )
  }

  return (
    <ScrollView
      ref={scrollRef}
      style={styles.waveformSamplesContainer}
      contentContainerStyle={s.alignItemsCenter}
      showsHorizontalScrollIndicator={false}
      scrollEnabled={false}
      horizontal
    >
      <View style={[s.row, s.alignItemsCenter]}>
        {samples.map(renderWaveform)}
      </View>
    </ScrollView>
  )
}

export default memo(AudioWaveformRecording)

const getStyles = createThemedStylesheet((theme) => {
  const styles = StyleSheet.create({
    hLine: {
      backgroundColor: theme.background.bg_2,
      height: 1,
      position: 'absolute',
      width: '100%',
      zIndex: 0,
    },
    waveformItem: {
      backgroundColor: theme.color.blue_400,
      borderRadius: UI_SIZE_4,
      marginHorizontal: WAVEFORM_SPACE,
      width: RECORDING_WAVEFORM_WIDTH,
    },
    waveformSamplesContainer: {
      marginRight: UI_SIZE_4,
    },
  })
  return styles
})
