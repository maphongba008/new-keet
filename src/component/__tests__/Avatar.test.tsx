import { render, screen } from '@testing-library/react-native'

import { Avatar } from 'component/Avatar'
import { APPIUM_IDs } from 'lib/appium'

jest.mock('../SvgIcon.tsx')

describe('Avatar component tests', () => {
  it('should render default avatar icon when no props are provided', () => {
    render(<Avatar />)
    const emptyIcon = screen.getByText('icon - keetOutline')
    expect(emptyIcon).toBeOnTheScreen()
  })
  it('should match snapshot and apply RGB color', () => {
    render(<Avatar color="#ff0000" />)
    expect(screen.toJSON()).toMatchSnapshot()
  })
  it('should match snapshot and apply named color', () => {
    const color = 'red'
    render(<Avatar color="red" />)
    expect(screen.toJSON()).toMatchSnapshot()
    const emptyIcon = screen.getByTestId(APPIUM_IDs.avatar_image)
    expect(JSON.stringify(emptyIcon.props.style)).toMatch(color)
  })
  it('should match snapshot and apply base64 icon', () => {
    const base64 =
      'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAAABHNCSVQICAg='
    render(<Avatar base64={base64} />)
    expect(screen.toJSON()).toMatchSnapshot()
    const img = screen.getByTestId(APPIUM_IDs.avatar_image)
    expect(img.props.source.uri).toMatch(base64)
  })
  it('should apply small style when small prop is provided', () => {
    const base64 =
      'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAAABHNCSVQICAg='
    render(<Avatar base64={base64} small />)
    expect(screen.toJSON()).toMatchSnapshot()
    const img = screen.getByTestId(APPIUM_IDs.avatar_image)
    expect(img).toHaveStyle({
      borderRadius: 15,
      height: 30,
      width: 30,
    })
  })
})
