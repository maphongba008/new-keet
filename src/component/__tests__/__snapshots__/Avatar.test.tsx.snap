// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar component tests should apply small style when small prop is provided 1`] = `
<Image
  accessibilityLabel="avatar_image"
  accessible={true}
  cachePolicy="memory"
  contentFit="contain"
  recyclingKey="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAAABHNCSVQICAg="
  source={
    {
      "uri": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAAABHNCSVQICAg=",
    }
  }
  style={
    {
      "borderRadius": 15,
      "height": 30,
      "width": 30,
    }
  }
  testID="avatar_image"
/>
`;

exports[`Avatar component tests should match snapshot and apply RGB color 1`] = `
<View
  accessibilityLabel="avatar_image"
  accessible={true}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#ff0000",
      "borderRadius": 36,
      "height": 72,
      "justifyContent": "center",
      "width": 72,
    }
  }
  testID="avatar_image"
>
  <Text>
    icon - 
    keetOutline
  </Text>
</View>
`;

exports[`Avatar component tests should match snapshot and apply base64 icon 1`] = `
<Image
  accessibilityLabel="avatar_image"
  accessible={true}
  cachePolicy="memory"
  contentFit="contain"
  recyclingKey="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAAABHNCSVQICAg="
  source={
    {
      "uri": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAAABHNCSVQICAg=",
    }
  }
  style={
    {
      "borderRadius": 36,
      "height": 72,
      "width": 72,
    }
  }
  testID="avatar_image"
/>
`;

exports[`Avatar component tests should match snapshot and apply named color 1`] = `
<View
  accessibilityLabel="avatar_image"
  accessible={true}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "red",
      "borderRadius": 36,
      "height": 72,
      "justifyContent": "center",
      "width": 72,
    }
  }
  testID="avatar_image"
>
  <Text>
    icon - 
    keetOutline
  </Text>
</View>
`;
